iron-fish/ironfish
# Add wallet/setAccountHead RPC

## Summary

Implements an RPC for manually updating an account's head. The user can pass in hashes of transactions known to be owned by the account, and the wallet will decrypt and apply only those transactions to the account's state.

## Needs Review

* [x] ~Should the name of the RPC route be different?~ Renamed from wallet/addKnownTransactions to wallet/setAccountHead
* [x] ~Should this logic live somewhere other than an RPC route, e.g. integrated with the wallet?~ We might move this out into the Scanner once it's refactored out of the wallet.
* [x] ~Should the route have wallet context instead of fullnode?~ Decided this is fine for now since it's similar to a service command anyway.
* [x] ~What should we do if someone cancels an RPC request before updateHead finishes?~ Added early-exit to the loops by checking `request.closed`.

## Testing Plan

Added automated tests for the RPC.

#### Manual testing

1. Run `wallet:scanning:off` to disable scanning for an account
1. Run `wallet:transactions` on the account. Save all transaction hashes.
1. Run `wallet:status`. Save the account's head hash as `end`
1. Run `wallet:rescan` to reset the account (#4985 will allow you to do this for one account). Cancel the scan
1. Run `wallet:status` again. Save the account's head hash as `start`
1. Run `start` on the directory.
1. In a separate prompt, run `repl` on the directory. Execute the command with the values you previously saved.
```
client = await sdk.connectRpc()
(await client.wallet.addKnownTransactions({ account: "account", start: "startHash", end: "endHash", transactions: [{ hash: "ceddfc8c0978db964f8fb3ad66f4d127235dcd25385401e2cd092a7c088c8992" }] })).content
```

After running `wallet:status` again, you should see the account restored to its former state.

## Documentation

Does this change require any updates to the Iron Fish Docs (ex. [the RPC API
Reference](https://ironfish.network/docs/onboarding/rpc/chain))? If yes, link a
related documentation pull request for the website.

```
[ ] Yes
```

## Breaking Change

Is this a breaking change? If yes, add notes below on why this is breaking and label it with `breaking-change-rpc` or `breaking-change-sdk`.

```
[ ] Yes
```
